src/context.rs

```diff
@@ -1170,6 +1174,47 @@ impl SeafowlContext for DefaultSeafowlContext {
                     None => location.into(),
                 };
 
+                // If the referenced table is in a cloud object store then register a new
+                // corresponding object store dynamically:
+                // 1. Using cmd.options if provided,
+                // 2. Otherwise use the object store credentials from the config if it matches
+                //    the object store kind.
+                let table_path = ListingTableUrl::parse(&cmd.location)?;
+                let scheme = table_path.scheme();
+                let url: &Url = table_path.as_ref();
+                if matches!(scheme, "s3" | "gs" | "gcs") {
+                    let bucket = url
+                        .host_str()
+                        .ok_or_else(|| {
+                            DataFusionError::Execution(format!(
+                                "Not able to parse bucket name from url: {}",
+                                url.as_str()
+                            ))
+                        })?
+                        .to_string();
+
+                    let config = if scheme == "s3" {
+                        let s3 = if cmd.options.is_empty() && let schema::ObjectStore::S3(s3) = self.internal_object_store.config.clone() {
+                            S3{ bucket, ..s3 }
+                        } else {
+                            S3::from_bucket_and_options(bucket, &cmd.options)?
```

Comment:
A potential alternative here is to use [get_s3_object_store_builder](https://docs.rs/datafusion-cli/latest/datafusion_cli/object_storage/fn.get_s3_object_store_builder.html) (and [get_s3_object_store_builder](https://docs.rs/datafusion-cli/latest/datafusion_cli/object_storage/fn.get_s3_object_store_builder.html) below, though that would mean adding a new dependency (datafusion-cli).

Reply:
Sorry I meant for the other link to point to [get_gcs_object_store_builder](https://docs.rs/datafusion-cli/latest/datafusion_cli/object_storage/fn.get_gcs_object_store_builder.html).

In the end both end up doing effectively what we want—[create an object store dynamically and register it](https://docs.rs/datafusion-cli/latest/src/datafusion_cli/exec.rs.html#200-227) with the object store registry.